In [1]:
pip install pandas

You should consider upgrading via the 'C:\Users\garca\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip' command.


In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('inadvance_synth.csv',sep=';')
print(data['age'].mean())

79.4033215326947


In [4]:
listnames = ['variable', 'missing']
df = data.isnull().sum()
new = pd.DataFrame(df)
new.columns =[ 'missing']

sorted_df = new.sort_values(by='missing')
#df.sort_values(, ascending = 'TRUE')

missing
Unnamed: 0                 0
estancias                  0
codservicioreal            0
label                      0
age                        0
metastatic_tumor          77
charlson                  77
numurgenciasprevias      552
drg                     3640
creatinina              7951
sodio                   8142
hematocrito             8302
leucocitos              8302
rdw_cv                  8334
rdw_sd                  8334
potasio                 8646
glucosa                 9228
codidiagingreso        11378
urea                   14260
proteina_c_reactiva    18175
num_grupoact3_HOSP     21001
barthel                28198

In [5]:
label = data['label']
counts = label.nunique()
label = pd.DataFrame(label)
label.columns =[ 'label']

label
0        0.0
1        0.0
2        0.0
3        0.0
4        0.0
...      ...
38411    0.0
38412    0.0
38413    0.0
38414    0.0
38415    0.0

[38416 rows x 1 columns]

In [6]:
label['label'].value_counts()

0.0    24985
1.0    13431
Name: label, dtype: int64

In [8]:
data['codservicioreal'] = data['codservicioreal'].astype('category')
data['codidiagingreso'] = data['codidiagingreso'].astype('category')
label['label']= label['label'].astype('category')
data['codservi2'] = data['codservicioreal'].cat.codes
data['codiagngr2'] = data['codidiagingreso'].cat.codes

In [10]:
y = label
X = data.drop(['Unnamed: 0','codidiagingreso','codservicioreal', 'label'], axis = 1) 

In [12]:
from pandas.api.types import is_numeric_dtype
def handleMissVar(df):
    for column in df:
        if(is_numeric_dtype(df[column])):
            mean = df[column].mean()
            df[column].fillna(mean,inplace=True)
    return df


In [13]:
X = handleMissVar(X)

In [14]:

from sklearn.model_selection import train_test_split
''' np.random.seed(100)
trainx, testx, trainy, testy = train_test_split(df, y, test_size=0.3, random_state = 100)
##test, train = np.split(df, [int(.3*len(df))])
training_set, validation_set = train_test_split(df,y test_size = 0.2, random_state = 21)
testy
'''

X_train, X_val,y_train,y_val = train_test_split(X,y, test_size = 0.2, random_state = 21)

In [16]:
X_train = handleMissVar(X_train)
X_val = handleMissVar(X_val)
y_train = handleMissVar(y_train)
y_val = handleMissVar(y_val)

1. Import the RandomForestClassifier and create the model with the default arguments

GridSearch

In [15]:
from sklearn.ensemble import RandomForestClassifier

In [23]:
from sklearn.model_selection import GridSearchCV
rf_params = {
    'max_depth' : [50, 150, 300],
    'n_estimators':[100,250, 350]
            }

RF = RandomForestClassifier(class_weight = 'balanced')
grid = GridSearchCV(RF, rf_params, cv=3, scoring='f1_weighted')
grid.fit(X_train, y2_train)
print(grid.best_params_)
print("Accuracy:"+ str(grid.best_score_))

C:\Users\garca\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\model_selection\_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
C:\Users\garca\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\model_selection\_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
C:\Users\garca\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\model_selection\_validation.py:680: DataConversionWarning: A column-vector y was passed

{'max_depth': 150, 'n_estimators': 350}
Accuracy:0.8220714859398646


****Model results using best Combination of Hyper-Params****

{'max_depth': 150, 'n_estimators': 350}
Accuracy:0.8220714859398646

In [24]:
RF = RandomForestClassifier(n_estimators= 350, max_depth = 150,class_weight = 'balanced')
RF.fit(X_train, y_train)
y_pred = RF.predict(X_val)

C:\Users\garca\AppData\Local\Temp/ipykernel_9564/1279577601.py:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  RF.fit(X_train, y_train)


In [18]:
from sklearn.metrics import accuracy_score,classification_report, precision_score, recall_score, roc_auc_score, f1_score,precision_recall_fscore_support


In [19]:
from sklearn.metrics import roc_curve, auc
import matplotlib.pyplot as plt
import random

Gradient Boosting

In [20]:
from sklearn.ensemble import GradientBoostingClassifier

**Gridsearch**

In [21]:
import numpy as np

In [1]:
y2_train = np.ravel(y_train) #para cambiar el formato a array que sino devuelve warning el gridsearch

NameError: name 'np' is not defined

In [26]:
from sklearn.model_selection import GridSearchCV
rf_params = {
    'loss': ['deviance','exponential'],
    'learning_rate' : [0.1, 0.3, 0.9],
    'n_estimators':[150, 300],
    'criterion' : ['friedman_mse', 'squared_error', 'mse', 'absolute_error']
            }

GB = GradientBoostingClassifier()
grid = GridSearchCV(GB, rf_params, cv=3, scoring='f1_weighted')
grid.fit(X_train, y2_train)
print(grid.best_params_)
print("Accuracy:"+ str(grid.best_score_))

C:\Users\garca\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\ensemble\_gb.py:471: FutureWarning: Criterion 'mse' was deprecated in v1.0 and will be removed in version 1.2. Use `criterion='squared_error'` which is equivalent.
  warnings.warn(
C:\Users\garca\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\tree\_classes.py:359: FutureWarning: Criterion 'mse' was deprecated in v1.0 and will be removed in version 1.2. Use `criterion='squared_error'` which is equivalent.
  warnings.warn(
C:\Users\garca\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\tree\_classes.py:359: FutureWarning: Criterion 'mse' was deprecated in v1.0 and will be removed in version 1.2. Use `criterion='squared_error'` which is equivalent.
  warnings.warn(
C:\Users\garca\AppDa

KeyboardInterrupt: 

Best combination of HP {'criterion': 'squared_error', 'learning_rate': 0.3, 'loss': 'deviance', 'n_estimators': 300}
Accuracy:0.9503072839361183

In [21]:
GB = GradientBoostingClassifier(loss = 'exponential', learning_rate = 0.3, n_estimators= 200, max_depth = 50)
GB.fit(X_train, y_train)
y_pred = GB.predict(X_val)

C:\Users\garca\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\ensemble\_gb.py:494: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVM

In [19]:
from sklearn import svm

In [20]:
from sklearn.model_selection import GridSearchCV
rf_params = {
    'C': [1, 10, 20],
    "kernel":['linear','poly','rbf','sigmoid']
            }

clf = svm.SVC( class_weight = 'balanced')
grid = GridSearchCV(clf, rf_params, cv=3, scoring='f1_weighted')
grid.fit(X_train, y2_train)
print(grid.best_params_)
print("Accuracy:"+ str(grid.best_score_))

C:\Users\garca\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Best combination of HyperParams

In [32]:
modl = svm.SVC(kernel = 'linear', C = 1000)
modl.fit(X_train, y_train)
y_pred = modl.predict(X_val)

C:\Users\garca\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [ ]:
false_positive_rate, true_positive_rate, thresholds = roc_curve(y_val, modl.predict_proba(X_val)[:,1])
roc_auc = auc(false_positive_rate, true_positive_rate)

print(f'AUC_ROC : ', roc_auc)
print(f"Accuray : {accuracy_score(y_val, y_pred) * 100:.2f}%")
print(f"Precision: {precision_score(y_val,y_pred, average='weighted')}")
print(f"Recall: {recall_score(y_val,y_pred, average='weighted')}")

plt.title('Label classification')
plt.plot(false_positive_rate, true_positive_rate, 'b',
label='AUC = %0.2f'% roc_auc)
plt.legend(loc='lower right')
plt.plot([0,1],[0,1],'r--')
plt.ylabel('True Positive Rate')
plt.xlabel('False Positive Rate')
plt.show()

Logistic Regression

In [18]:
from sklearn.linear_model import LogisticRegression

In [19]:
from sklearn.model_selection import GridSearchCV
rf_params = {
    'C': [1,10, 20],
    'penalty': ['l1', 'l2', 'elasticnet', 'none'],
    "solver":['newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga']
            }

lnr = LogisticRegression(class_weight = 'balanced')
grid = GridSearchCV(lnr, rf_params, cv=3, scoring='f1_weighted')
grid.fit(X_train, y2_train)
print(grid.best_params_)
print("Accuracy:"+ str(grid.best_score_))

C:\Users\garca\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\garca\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\garca\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y t

{'C': 1, 'penalty': 'l2', 'solver': 'liblinear'}
Accuracy:0.7612973988077346


Best combination
{'C': 1, 'penalty': 'l2', 'solver': 'liblinear'}
Accuracy:0.7612973988077346